internal/postgres/process_test.go

```diff
@@ -0,0 +1,589 @@
+		},
+		{
+			name: "INSERT with spaces",
+			input: "CREATE TABLE test (a text NOT NULL, b text NOT NULL, n bigint);\n" +
+				"ALTER TABLE ONLY test ADD CONSTRAINT test_pkey PRIMARY KEY (a, b);\n" +
+				"INSERT INTO test (a, b, n) VALUES (' a42 ', '\nb6\n', 2);\n",
+			expectedData: []spannerData{
+				spannerData{table: "test", cols: []string{"a", "b", "n"}, vals: []interface{}{" a42 ", "\nb6\n", int64(2)}}},
+		},
+		{
+			name: "Statements with embedded semicolons",
+			input: "CREATE TABLE test (a text NOT NULL, b text NOT NULL, n bigint);\n" +
+				"ALTER TABLE ONLY test ADD CONSTRAINT test_pkey PRIMARY KEY (a, b);\n" +
+				"INSERT INTO test (a, b, n) VALUES ('a;\n2', 'b;6\n', 2);\n",
+			expectedData: []spannerData{
+				spannerData{table: "test", cols: []string{"a", "b", "n"}, vals: []interface{}{"a;\n2", "b;6\n", int64(2)}}},
+			expectIssues: true,
+		},
+		{
+			name: "Tables and columns with illegal characters",
+			input: `CREATE TABLE "te.s t" ("a?^" text PRIMARY KEY, "b.b" text, "n*n" bigint);
+                                INSERT INTO "te.s t" ("a?^", "b.b", "n*n") VALUES ('a', 'b', 2);`,
+			expectedData: []spannerData{
+				spannerData{table: "te_s_t", cols: []string{"a__", "b_b", "n_n"}, vals: []interface{}{"a", "b", int64(2)}}},
+			expectedSchema: map[string]ddl.CreateTable{
+				"te_s_t": ddl.CreateTable{
+					Name: "te_s_t",
+					Cols: []string{"a__", "b_b", "n_n"},
+					Cds: map[string]ddl.ColumnDef{
+						"a__": ddl.ColumnDef{Name: "a__", T: ddl.String{Len: ddl.MaxLength{}}, NotNull: true},
+						"b_b": ddl.ColumnDef{Name: "b_b", T: ddl.String{Len: ddl.MaxLength{}}},
+						"n_n": ddl.ColumnDef{Name: "n_n", T: ddl.Int64{}},
+					},
+					Pks: []ddl.IndexKey{ddl.IndexKey{Col: "a__"}}}},
+		},
+		{
+			name: "Tables and columns with illegal characters: CREATE-ALTER",
+			input: `CREATE TABLE "te.s t" ("a?^" text, "b.b" text, "n*n" bigint);
+				ALTER TABLE ONLY "te.s t" ADD CONSTRAINT test_pkey PRIMARY KEY ("a?^");
+                                INSERT INTO "te.s t" ("a?^", "b.b", "n*n") VALUES ('a', 'b', 2);`,
+			expectedData: []spannerData{
+				spannerData{table: "te_s_t", cols: []string{"a__", "b_b", "n_n"}, vals: []interface{}{"a", "b", int64(2)}}},
+			expectedSchema: map[string]ddl.CreateTable{
+				"te_s_t": ddl.CreateTable{
+					Name: "te_s_t",
+					Cols: []string{"a__", "b_b", "n_n"},
+					Cds: map[string]ddl.ColumnDef{
+						"a__": ddl.ColumnDef{Name: "a__", T: ddl.String{Len: ddl.MaxLength{}}, NotNull: true},
+						"b_b": ddl.ColumnDef{Name: "b_b", T: ddl.String{Len: ddl.MaxLength{}}},
+						"n_n": ddl.ColumnDef{Name: "n_n", T: ddl.Int64{}},
+					},
+					Pks: []ddl.IndexKey{ddl.IndexKey{Col: "a__"}}}},
+		},
+		// The "Data conversion: ..." cases check data conversion for each type.
+		{
+			name: "Data conversion: bool, bigserial, char, bytea",
+			input: `
+CREATE TABLE test (id integer PRIMARY KEY, a bool, b bigserial, c char, d bytea);
```

Comment:
accidental newline? other places as well.

Reply:
Actually I formatted the COPY-FROM tests like this on purpose.

The reason is that the copy-from block (the bit with the data that is ended with \.) has to be very specifically formatted. In particular, it has to use \t to separate data elements, and any whitespace is significant. So this part was easier to format using raw string literals. Also, its easier to cut-and-paste examples from pg_dump files.

I added an extra new line at the start so everything would line up -- extra whitespace is fine in CREATE TABLE statements.

But I'm open to changing this?

Reply:
that makes sense. You can leave it as is.